### PR TITLE
[SPARK-12048][SQL] Part 2 Prevent to close JDBC resources twice

### PR DESCRIPTION
**Pull request by tedyu**
PR #10101 dealt with sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala

This PR applies the same technique on JdbcRDD.scala

### CODE REVIEW

**Comment by tedyu** [comment]
@andrewor14 @zsxwing 
Please take a look

File: core/src/main/scala/org/apache/spark/rdd/JdbcRDD.scala [hunk 1]
```diff
@@ -100,6 +101,7 @@ class JdbcRDD[T: ClassTag](
     }
 
     override def close() {
+      if (closed) return
       try {
         if (null != rs) {
           rs.close()
```
Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/rdd/JdbcRDD.scala]: From my understanding, we could call `rs = null` after it is closed, also for `stmt` and `conn` in the same way to make this function is reenterable.

Comment (srowen) [inline core/src/main/scala/org/apache/spark/rdd/JdbcRDD.scala]: That's another approach, though the other implementation opted for a close flag

**Comment by zsxwing** [comment]
I don't think we need this patch. I saw the following comments in javadoc

```
     * Calling the method <code>close</code> on a <code>Connection</code>
     * object that is already closed is a no-op.

     * Calling the method <code>close</code> on a <code>Statement</code>
     * object that is already closed has no effect.

     * Calling the method <code>close</code> on a <code>ResultSet</code>
     * object that is already closed is a no-op.
```

**Comment by srowen** [comment]
Javadoc of which? the JDK interfaces? I think the issue was that SQLite didn't actually seem to like being closed twice in practice.

**Comment by zsxwing** [comment]
> Javadoc of which? the JDK interfaces? I think the issue was that SQLite didn't actually seem to like being closed twice in practice.

Forgot to read the JIRA at first. Didn't know that `SQLite` doesn't follow the Javadoc of the JDK interfaces.

**Comment by tedyu** [comment]
I compared JDBCRDD.scala with JdbcRDD.scala
From what I can tell according to the usage of java.sql.Connection and java.sql.ResultSet, the proposed change is needed

**Comment by tedyu** [comment]
@zsxwing @srowen 
Anything I need to do for this PR ?

Thanks

**Comment by zsxwing** [comment]
Looks `NextIterator` already has a `closed` field.

**Comment by tedyu** [comment]
Looks like the case is covered by NextIterator already

**Pull request closed without merging**